### DISCUSSION TITLE
Feedback on SPEC.md

### DISCUSSION THREAD

**Discussion by gunnarmorling** [recap-build/recap#222] [General]
Hey @criccomini, a few random thoughts:

* The "forward references" were confusing me at first; e.g. you make use of `int32` before actually introducing this type. An explainer upon the first reference would be good ("short-form alias for an int with 32 bits, see ... for details")
* Attributes of list and map elements seems inconsistent: list explicitly mentions `bits`, how would a `list[string(bytes=100]` be defined? in contrast, no attributes of that kind are mentioned for map keys/values; also `signed` is mentioned for list in the example, but not the definition; I think this is mostly down to how types are represented in definitions and examples consistently
* Field: for default you mention type literal, whereas later in FAQ you say there is no literal type; I think for default values, the different literal types need spelling out
* Units for temporal aliases: could be enums?

**Comment by criccomini** [comment]
Thx! I'll update the spec accordingly. I'm pressure testing the spec against a Python implementation right now, so it might take me a bit to push the changes.

**Comment by criccomini** [comment]
> The "forward references" were confusing me at first; e.g. you make use of int32 before actually introducing this type. An explainer upon the first reference would be good ("short-form alias for an int with 32 bits, see ... for details")

I can see why this is confusing. The attribute types are NOT the same as Recap types.

The reason I used `int32` for the `bytes` attribute, for example, is to express a min/max allowed value for `bytes`. A lot of specs (e.g. Avro) just say `int` for attributes like `bytes`. In those specs, it's not clear whether the `bytes`'s implementation is supposed to use a signed or unsigned int, and whether it's supposed to use 32-bit or 64-bit (or unlimited?) integers. This actually matters in the implementation because it's an implicit constraint: if you store the `bytes` attribute in a 32-bit signed integer, the maximum number of bytes you can have in the field is 2_147_483_647.

All that said, I will clarify the spec to make this more obvious.

> Attributes of list and map elements seems inconsistent: list explicitly mentions bits, how would a list[string(bytes=100] be defined? in contrast, no attributes of that kind are mentioned for map keys/values; also signed is mentioned for list in the example, but not the definition; I think this is mostly down to how types are represented in definitions and examples consistently

So, the `bits` for `list` was meant to capture the leading integer when encoding an array. Again, this is a constraint thing. If `bits` is set to 32, then the maximum number of elements in the list would be INT_MAX (or UINT_MAX--I didn't clarify this in the spec). Basically, same deal as the `bytes` example above.

That said, I suspect the better way to model a list is with the following attributes:

* `value`: The type of the items in the list (required).
* `length`: An optional `int32` defining the maximum length of the list (similar to `bytes` for byte/string).
* `variable`: A required `bool` (default true) defining whether the list is variable length or not.

Thus, a LinkedList (with unlimited length) would have `variable=true` and `length` unset. A fixed-size array of 100 elements would have `length=100` and `variable=false`.

I didn't include the `bits` field for maps because I haven't really come across fixed-size or max-size dictionaries/maps. Do you think this is something we should add? Most specs encode maps without a fixed max size (like an int32 or int64). For example, Avro says, "Maps are encoded as a series of blocks. Each block consists of a long count value, followed by that many key/value pairs. A block with count zero indicates the end of the map. Each item is encoded per the map's value schema."

> Field: for default you mention type literal, whereas later in FAQ you say there is no literal type; I think for default values, the different literal types need spelling out

Yea, this is a tricky one. So, Recap's type system doesn't have literals. But again, similar to the first comment, the *attribute* type for `default` is a literal. This is similar to the way [Avro behaves](https://avro.apache.org/docs/1.10.2/spec.html#schema_record). I think the confusion is mostly because I was very careless about distinguishing between Recap types and attribute types.

> Units for temporal aliases: could be enums?

Yep. I'll add this.

**Reply by gunnarmorling** [reply]
> The attribute types are NOT the same as Recap types

Oooh, I see. That was not clear indeed. Are the attribute types documented at all? It feels a bit irregular to me that there is a separate "type system" for them. Why not using the Recap types there?

**Reply by gunnarmorling** [reply]
> I didn't include the bits field for maps because I haven't really come across fixed-size or max-size dictionaries/maps. Do you think this is something we should add

No, I was just confused by the way all this was represented. It makes sense now after the update.

**Comment by criccomini** [comment]
I've updated the spec accordingly:

https://github.com/recap-cloud/recap/blob/main/SPEC.md
